genenetwork/genenetwork2
# Build/python3 migration

#### Description

This PR:
- Migrates python2 to python3 using the tool `2to3.8`
- Fixes failing tests(Mechanical Rob and unittests) along the way
- Removes unused imports in some cases
- Applies pep8 in some cases(like replacing tabs with 4 spaces)
- Replaces python2-htmlgen to python3-htmlgen

#### How should this be tested?

- Install python3-genenetwork from guix(see below). This has all the python3 packages you need.
- Confirm that all unittests pass by running:

```bash
# Replace paths where appropriate
env GN2_PROFILE=/home/bonface/opt/python3-genenetwork2 TMPDIR=/home/bonface/tmp SERVER_PORT=5004 GENENETWORK_FILES=/home/bonface/gnu/data/gn2_data/ /home/bonface/projects/genenetwork2/bin/genenetwork2 /home/bonface/projects/genenetwork2/etc/default_settings.py -c -m unittest discover -v
```

- Confirm that Mechanical Rob works:

```bash
# Replace paths and host where appropriate
time env GN2_PROFILE=/home/bonface/opt/genenetwork2 TMPDIR=/home/bonface/tmp SERVER_PORT=5004 GENENETWORK_FILES=/home/bonface/gnu/data/gn2_data/ ./bin/genenetwork2 ./etc/default_settings.py -c /home/bonface/projects/genenetwork2/test/requests/test-website.py -a http://localhost:5004
```

- Run the site(the proxy has to be up) and check that things work as expected:
```
cd /home/bonface/projects/genenetwork2 && env SERVER_PORT=5004 TMPDIR=/home/bonface/tmp WEBSERVER_MODE=DEBUG LOG_LEVEL=DEBUG GENENETWORK_FILES=/home/bonface/gnu/data/gn2_data/
GN2_PROFILE=/home/bonface/opt/python3-genenetwork2 /home/bonface/projects/genenetwork2/scripts/run_debug.sh
```

#### Any background context you want to provide?

- *gunicorn* was not working for me(still don't know why), so I used a version provided by archlinux.
- Right now, things just work. I don't know if they work correctly though. There will be some bugs/ errors that me and @zach will iteratively have to work on. But we have something that works!
- I caught some bugs through the few existing unittests that evaded 2to3.8 like https://github.com/genenetwork/genenetwork2/pull/422/commits/7717f0d3d4802d115a8b5ace4d379864ba5b1188. I'm sure there are some subtle bugs lurking around, so we'd have to be wary of those.

#### What are the relevant pivotal tracker stories?

- fixes genenetwork/genenetwork2#424
- fixes genenetwork/genenetwork2#398

#### Screenshots (if appropriate)

N/A

#### Questions

- I've had to do some hacks to make the auth stuff work:
  - https://github.com/genenetwork/genenetwork2/pull/422/commits/70dbeeb5832711ed5271434e482c18bc7ea095b8
  - https://github.com/genenetwork/genenetwork2/pull/422/commits/b46faf479e129801f4bec2457444b5341e8aeabf
  - https://github.com/genenetwork/genenetwork2/pull/422/commits/46443ec8d2cdfd7c60358a889d90a90e4f7daaf4
  - https://github.com/genenetwork/genenetwork2/pull/422/commits/7f623509a4afded47eb8580451c146487d2ef662

- Also, we have to check if the utf-8 bugs still exist. Python3 handles utf-8 strings a bit differently. 

Could @zsloan have a look at the above and give me some feedback?

#### How to install python3-genenetwork from guix

Recommended: install python3-genenetwork in its own profile. In this case, we have guix-bioinformatics and guix-past set up locally. You could always add this to it's own *channel.scm*` file.

```bash
env GUIX_PACKAGE_PATH="/home/bonface/projects/guix-bioinformatics:/home/bonface/projects/guix-past/modules" guix install python3-genenetwork2 -p ~/opt/python3-genenetwork2

```
